### ISSUE TITLE
How to fixed partial parameters and only train part of the model?

### ISSUE THREAD

**Issue by greeeenmouth** [open-mmlab/mmocr#1032]
Hi,
I want to finetune the decoder of my model. How to fixed partial parameters of the specific part such as backbone, encoder and only train the decoder part?

I have seen the issue of [https://github.com/open-mmlab/mmocr/issues/828](url), but it will lead to another question [https://github.com/open-mmlab/mmocr/issues/958](url). Also, I'm a little confused about my question and whether it's the same with [https://github.com/open-mmlab/mmocr/issues/828](url).

**Comment by Mountchicken** [comment]
Hello there @greeeenmouth 
- I'm not sure if the learning rate problem in [PR958](https://github.com/open-mmlab/mmocr/issues/958) is fixed. @gaotongxiao 
- However, I have a less elegant solution. I guess you are trying to fine-tune a recognition model as you mentioned encoder and decoder. 
- Most recognizers are inheritd from [EncodeDecodeRecognizer](https://github.com/open-mmlab/mmocr/blob/main/mmocr/models/textrecog/recognizer/encode_decode_recognizer.py#L13), except for `ABINet`. And you can freeze the parameter here.
https://github.com/open-mmlab/mmocr/blob/86879c683416d7c000d479e9216602166c638a33/mmocr/models/textrecog/recognizer/encode_decode_recognizer.py#L42-L56
- Let's say you want to freeze backbone and encoder, and you modify the code as follow:
```python
assert backbone is not None
self.backbone = build_backbone(backbone)
for p in self.backbone.parameters():
    p.requires_grad = False

# Encoder module
self.encoder = None
if encoder is not None:
    self.encoder = build_encoder(encoder)
    for p in self.encoder.parameters():
       p.requires_grad = False
```

**Comment by gaotongxiao** [comment]
https://github.com/open-mmlab/mmocr/issues/958 has been added to MMCV's backlog but it might take time to solve. BTW, #828 still fits your situation.

**Comment by greeeenmouth** [comment]
That's exactly what I need. Thank you so much!

**Comment by simplify23** [comment]
I tried to freeze part of the model for training and this problem helped me a lot. But I'm still a little confused.

I found that `custom_keys` can customize the freeze model in the issue  #828 . However in the log `the loss converges much slower than that without freezing the backbone and the neck`. Freezing the model theoretically leads to slower convergence (and I think there is also a performance loss compared to end-to-end training). Is the MMOCR learning rate of the exception just an error on the logging display, or it also has an impact on convergence or performance?

Besides, @Mountchicken 
Similarly, does the idea of `p.requires_grad = False` descent lead to performance degradation (in addition to the freezing of the model)

**Comment by doem97** [comment]
> I tried to freeze part of the model for training and this problem helped me a lot. But I'm still a little confused.
> 
> I found that `custom_keys` can customize the freeze model in the issue #828 . However in the log `the loss converges much slower than that without freezing the backbone and the neck`. Freezing the model theoretically leads to slower convergence (and I think there is also a performance loss compared to end-to-end training). Is the MMOCR learning rate of the exception just an error on the logging display, or it also has an impact on convergence or performance?
> 
> Besides, @Mountchicken Similarly, does the idea of `p.requires_grad = False` descent lead to performance degradation (in addition to the freezing of the model)

Hi, have you solved this? Linear probing (only train decoder) truly leads to worse performance, as the encoder is not well tuned for the recognizor's decoder. However, by far my convergence is not too slow, cause we only need to train the decoder. May refer to "linear probing" results in SeqCLR or MaskOCR.
